flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosApplicationMasterRunner.java

```diff
@@ -0,0 +1,618 @@
+			LOG.debug("Mesos dynamic properties: {}", dynamicProperties);
+
+			final Configuration config = createConfiguration(workingDir, dynamicProperties);
+
+			// Mesos configuration
+			final MesosConfiguration mesosConfig = createMesosConfig(config, appMasterHostname);
+
+			// environment values related to TM
+			final int taskManagerContainerMemory;
+			final int numInitialTaskManagers;
+			final int slotsPerTaskManager;
+
+			try {
+				taskManagerContainerMemory = Integer.parseInt(ENV.get(MesosConfigKeys.ENV_TM_MEMORY));
+			} catch (NumberFormatException e) {
+				throw new RuntimeException("Invalid value for " + MesosConfigKeys.ENV_TM_MEMORY + " : "
+					+ e.getMessage());
+			}
+			try {
+				numInitialTaskManagers = Integer.parseInt(ENV.get(MesosConfigKeys.ENV_TM_COUNT));
+			} catch (NumberFormatException e) {
+				throw new RuntimeException("Invalid value for " + MesosConfigKeys.ENV_TM_COUNT + " : "
+					+ e.getMessage());
+			}
+			try {
+				slotsPerTaskManager = Integer.parseInt(ENV.get(MesosConfigKeys.ENV_SLOTS));
+			} catch (NumberFormatException e) {
+				throw new RuntimeException("Invalid value for " + MesosConfigKeys.ENV_SLOTS + " : "
+					+ e.getMessage());
+			}
+
+			final ContaineredTaskManagerParameters containeredParameters =
+				ContaineredTaskManagerParameters.create(config, taskManagerContainerMemory, slotsPerTaskManager);
+
+			final MesosTaskManagerParameters taskManagerParameters =
+				MesosTaskManagerParameters.create(config, containeredParameters);
+
+			LOG.info("TaskManagers will be created with {} task slots",
+				taskManagerParameters.containeredParameters().numSlots());
+			LOG.info("TaskManagers will be started with container size {} MB, JVM heap size {} MB, " +
+					"JVM direct memory limit {} MB, {} cpus",
+				taskManagerParameters.containeredParameters().taskManagerTotalMemoryMB(),
+				taskManagerParameters.containeredParameters().taskManagerHeapSizeMB(),
+				taskManagerParameters.containeredParameters().taskManagerDirectMemoryLimitMB(),
+				taskManagerParameters.cpus());
+
+			// JM endpoint, which should be explicitly configured by the dispatcher (based on acquired net resources)
+			final int listeningPort = config.getInteger(ConfigConstants.JOB_MANAGER_IPC_PORT_KEY,
+				ConfigConstants.DEFAULT_JOB_MANAGER_IPC_PORT);
+			require(listeningPort >= 0 && listeningPort <= 65536, "Config parameter \"" +
+				ConfigConstants.JOB_MANAGER_IPC_PORT_KEY + "\" is invalid, it must be between 0 and 65536");
+
+			// ----------------- (2) start the actor system -------------------
+
+			// try to start the actor system, JobManager and JobManager actor system
+			// using the configured address and ports
+			actorSystem = BootstrapTools.startActorSystem(config, appMasterHostname, listeningPort, LOG);
+
+			final String akkaHostname = AkkaUtils.getAddress(actorSystem).host().get();
+			final int akkaPort = (Integer) AkkaUtils.getAddress(actorSystem).port().get();
```

Comment:
`AkkaUtils.getAddress(...)` could be stored. Then we don't have to recreate the `RemoteAddressExtension`.